### DISCUSSION TITLE
M.2 B+M with a USB Adaptor. Possible?

### DISCUSSION THREAD

**Discussion by lolouk44** [blakeblackshear/frigate#1916] [General]
Since there's not coral USB available anywhere for the next ~6 months, and since the M.2 is less than half the price, I've purchased the M.2 B+M Coral ([this one](https://thepihut.com/products/google-coral-m-2-accelerator-b-m-key))
I'm hoping I can use it with either a USB adapter like [this one](https://www.ebay.co.uk/itm/144185813728?hash=item219224eae0:g:tNUAAOSwqSxhM7tc), or with a M.2 to PCIE X4 like [this one](https://www.ebay.co.uk/itm/363056412446?hash=item5487d9371e:g:vhQAAOSw5iJgN46L).
Has anbody done this successfully?

**Comment by shred86** [comment]
There's some USB Coral Edge TPU's available here: https://www.arrow.com/en/products/g950-06809-01/google-corporation?q=G950-06809-01

Unknown on if you can use the M.2 version with a USB adapter. I'm curious as well!

**Comment by lolouk44** [comment]
Well [this comment](https://community.home-assistant.io/t/pcie-adapter-for-google-coral/339521/3) suggest it'll work so I've ordered the PCIe adapter. Let's see...
I'll post an update when I get all the parts.

**Reply by jefftidwell-lab** [reply]
I also just ordered the PCIE one[ (this one)](https://thepihut.com/products/google-coral-m-2-accelerator-b-m-key), I have a few m.2 adapters and PCIE adapters, Should be here Saturday or Monday. I'm running on an intel NUC i3, which has a m.2 slot, but I'm using it for the system storage.

**Reply by Davy-H** [reply]
I've got them working successfully using M.2 to PCie adapters (using PCIe 1x Gen2). Got a couple of these cheaper adapters off aliexpress: https://www.aliexpress.com/item/4000212106479.html?spm=a2g0s.9042311.0.0.1f304c4dvx44n6 

![image](https://user-images.githubusercontent.com/89611653/145471878-ff1e5327-889c-45da-9a63-fc9a612fda91.png)

They work great for me running frigate in a docker container.

**Comment by lolouk44** [comment]
Well it's supposed to work, at least when following [these instructions](https://coral.ai/docs/m2/get-started/#requirements)
My PC is PCI Express 2.0 so I'm screwed...
 will fork out for a USB Coral...

**Reply by tortho** [reply]
Are these instructions something we have to do and is possible to do on HassOS?
https://coral.ai/docs/m2/get-started/#requirements

**Comment by chokays** [comment]
![NCQnqhECdy1xQnsxfVEFB5vy](https://user-images.githubusercontent.com/95276985/145470521-8aad379a-8ccb-47e8-8602-2a6618a4ce62.png)
Don't think an USB adapter will work. The adapters I found only work for storage devices.

I ended up buying the only TPU type in stock in my region and using an M.2 adapter to convert to from AE key to BM

**Comment by lolouk44** [comment]
Yeah that would work but don't make the same mistake as me,  you need to make sure your motherboard supports PCIE 3.0 (my old one is PCIE 2.0)

**Reply by Davy-H** [reply]
What issues are you experiencing? They work great for me at with Gen 2 (5Gbps).

04:00.0 System peripheral: Global Unichip Corp. Coral Edge TPU (prog-if ff)
	Subsystem: Global Unichip Corp. Coral Edge TPU
	Control: I/O- Mem+ BusMaster+ SpecCycle- MemWINV- VGASnoop- ParErr- Stepping- SERR- FastB2B- DisINTx+
	Status: Cap+ 66MHz- UDF- FastB2B- ParErr- DEVSEL=fast >TAbort- <TAbort- <MAbort- >SERR- <PERR- INTx-
	Latency: 0, Cache Line Size: 64 bytes
	Interrupt: pin A routed to IRQ 19
	Region 0: Memory at f0100000 (64-bit, prefetchable) [size=16K]
	Region 2: Memory at f0000000 (64-bit, prefetchable) [size=1M]
	Capabilities: [80] Express (v2) Endpoint, MSI 00
		DevCap:	MaxPayload 256 bytes, PhantFunc 0, Latency L0s <64ns, L1 <1us
			ExtTag+ AttnBtn- AttnInd- PwrInd- RBE+ FLReset- SlotPowerLimit 10.000W
		DevCtl:	CorrErr- NonFatalErr- FatalErr- UnsupReq-
			RlxdOrd- ExtTag+ PhantFunc- AuxPwr- NoSnoop+
			MaxPayload 128 bytes, MaxReadReq 512 bytes
		DevSta:	CorrErr- NonFatalErr- FatalErr- UnsupReq- AuxPwr- TransPend-
		LnkCap:	Port #1, Speed 5GT/s, Width x1, ASPM L0s L1, Exit Latency L0s <64ns, L1 <1us
			ClockPM+ Surprise- LLActRep- BwNot- ASPMOptComp+
		LnkCtl:	ASPM Disabled; RCB 64 bytes Disabled- CommClk-
			ExtSynch- ClockPM+ AutWidDis- BWInt- AutBWInt-
		LnkSta:	Speed 5GT/s (ok), Width x1 (ok)

**Reply by lolouk44** [reply]
It just doesn't get recognised for me. I tried the coral drivers and it just fails to get recognised. How did your install the drivers?

**Reply by Davy-H** [reply]
All I did was this
  echo "deb https://packages.cloud.google.com/apt coral-edgetpu-stable main" | sudo tee /etc/apt/sources.list.d/coral-edgetpu.list
  curl https://packages.cloud.google.com/apt/doc/apt-key.gpg | sudo apt-key add -
  sudo apt-get update
  sudo apt-get install gasket-dkms libedgetpu1-std
  sudo groupadd apex
  sudo adduser $USER apex

But even without this you should be able to do
lspci -nn | grep 089a

And see the coral devices.

**Reply by tortho** [reply]
None if these commands seems to be possible if you run HomeAssistantOS

**Reply by lolouk44** [reply]
@Davy-H Sorry for the late reply. Had to get the PC open from its hiding place and wanted to wait until I had received a spare SSD I ordered.
Not sure what's different between what I had done and what you suggested above, but I can confirm that I have a working PCI edgetpu. Thanks a lot :)

**Reply by Davy-H** [reply]
Awesome! Glad to hear you got it working :)
